src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,218 @@
+            }
+
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _options = options;
+            _logger = loggerFactory.CreateLogger<HttpLoggingMiddleware>();
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>HttpResponseLog.cs
+        public async Task Invoke(HttpContext context)
+        {
+            var options = _options.CurrentValue;
+            RequestBufferingStream? requestBufferingStream = null;
+            Stream? originalBody = null;
+            if ((HttpLoggingFields.Request & options.LoggingFields) != HttpLoggingFields.None)
+            {
+                var request = context.Request;
+                var list = new List<KeyValuePair<string, object?>>(
+                    request.Headers.Count + DefaultRequestFieldsMinusHeaders);
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.Protocol))
+                {
+                    AddToList(list, nameof(request.Protocol), request.Protocol);
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.Method))
+                {
+                    AddToList(list, nameof(request.Method), request.Method);
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.Scheme))
+                {
+                    AddToList(list, nameof(request.Scheme), request.Scheme);
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.Path))
+                {
+                    AddToList(list, nameof(request.PathBase), request.PathBase);
+                    AddToList(list, nameof(request.Path), request.Path);
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.Query))
+                {
+                    AddToList(list, nameof(request.QueryString), request.QueryString.Value);
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestHeaders))
+                {
+                    FilterHeaders(list, request.Headers, options.AllowedRequestHeaders);
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestBody)
+                    && MediaTypeHelpers.TryGetEncodingForMediaType(request.ContentType,
+                        options.SupportedMediaTypes,
+                        out var encoding))
+                {
+                    originalBody = request.Body;
+                    requestBufferingStream = new RequestBufferingStream(
+                        request.Body,
+                        options.RequestBodyLogLimit,
+                        _logger,
+                        encoding);
+                    request.Body = requestBufferingStream;
+                }
```

Comment:
// TODO: connection logging for bodies without a content-type?

Reply:
https://github.com/dotnet/aspnetcore/issues/31884